src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -142,11 +176,18 @@ public async Task<bool> OnDocumentOpenedAsync(TextDocument document)
 
             public async Task<bool> OnDocumentClosedAsync(TextDocument document, IGlobalOptionService globalOptions)
             {
-                // can not be cancelled
-                // remove active file state and put it in project state
-                if (!_activeFileStates.TryRemove(document.Id, out var activeFileState))
+                ActiveFileState? activeFileState;
+
+                lock (_lock)
```

Comment:
Note that `OnDocumentClosedAsync` is only called from solution crawler code path, and is guaranteed to never be invoked concurrently. That is the reason prior code works fine without any locking. I gave some more context offline about the partial thread-safe nature of these types in presence of both LSP and non-LSP worlds. Hopefully, we can clean up these types completely once we move away from solution crawler and non-LSP world.